src/server/scene/surface_stack.cpp

```diff
@@ -425,6 +425,110 @@ void ms::SurfaceStack::raise(SurfaceSet const& ss)
     }
 }
 
+void ms::SurfaceStack::swap_z_order(SurfaceSet const& first, SurfaceSet const& second)
+{
+    {
+        RecursiveWriteLock ul(guard);
+        for (auto& layer : surface_layers)
+        {
+            // The goal is to swap the first set with the second set such that their Z-order is swapped.
+
+            // First, find the start and end of the range that we want to swap
+            auto swap_begin = layer.begin();
+            auto swap_end = layer.end();
+            size_t num_first_found = 0;
+            size_t num_second_found = 0;
+            bool first_to_front = false;
+            for (auto it = layer.begin(); it != layer.end(); it++)
+            {
+                // Once we've found them all, then we have the whole range to sort across
+                if (num_first_found == first.size() && num_second_found == second.size())
+                {
+                    swap_end = it;
+                    break;
+                }
+
+                // Find the start position and count how many we've found
+                bool in_first = first.count(*it) > 0;
+                bool in_second = second.contains(*it) > 0;
+                if (in_first && in_second)
+                {
+                    BOOST_THROW_EXCEPTION(std::logic_error("Swapped sets must not have any overlap"));
+                }
```

Comment:
Mixing precondition checks with logic makes for hard to read code.

If we need to validate preconditions (which is debatable here), then it is best at the start of the function:

```c++
    for (auto const& s : first)
    {
        if (second.count(s))
            fatal_error("Swapped sets must not have any overlap");
    }
```

And, if anywhere, this check belongs in `BasicWindowManager` as that is where we can first detect a problem. (In that context, it is clear that `first` and `second` are the same `SurfaceSet`, so a no-op would also be sensible behaviour)